### PR TITLE
[WIP] port cache plugin internals to pathlib

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
warning logging got broken by detanglement from config

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](/changelog/README.rst) for details.
- [ ] Include new tests or update existing tests when applicable.

### CODE REVIEW

File: src/_pytest/cacheprovider.py [hunk 1]
```diff
@@ -5,39 +5,41 @@
...
 import six
+import attr
 
 import pytest
 import json
-import os
 from os.path import sep as _sep, altsep as _altsep
+import shutil
+
+from . import paths
```
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py]: I think you forgot to include `paths.py` into the commit.

File: src/_pytest/cacheprovider.py [hunk 2]
```diff
@@ -5,39 +5,41 @@
...
-            self._cachedir.mkdir()
+    @classmethod
+    def for_config(cls, config):
+        cachedir = cls.cache_dir_from_config(config)
+        if config.getoption("cacheclear") and cachedir.exists():
+            shutil.rmtree(str(cachedir))
+            cachedir.mkdir()
+        return cls(cachedir)
+
+    _cachedir = attr.ib(repr=False)
```
Comment (nicoddemus) [inline src/_pytest/cacheprovider.py]: For consistency we maybe should keep all `attr.ib` declarations at the start of the class declaration, before any methods?

File: src/_pytest/compat.py [hunk 3]
```diff
@@ -29,8 +29,11 @@
 
 if _PY3:
     from inspect import signature, Parameter as Parameter
+    from pathlib import Path
 else:
     from funcsigs import signature, Parameter as Parameter
+    from pathlib2 import Path
+Path
```
Comment (nicoddemus) [inline src/_pytest/compat.py]: Extra symbol here, accident?

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt it seems there's one failing test still.

File: src/_pytest/cacheprovider.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/cacheprovider.py]: fixed

**Comment by RonnyPfannschmidt** [comment]
man. py34 is broken due to its broken pathlib ^^ part of me wants to drop support, will fix later

**Comment by nicoddemus** [comment]
Do we have an idea of when we plan to drop py34 btw?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus https://devguide.python.org/#status-of-python-branches - eol is in about 9 months, i#d wait at least another 6 months

**Comment by nicoddemus** [comment]
Fair enough, thanks for the link!

**Comment by RonnyPfannschmidt** [comment]
seems i was wrong about 3.4, needs 3.5 as well i suppose

File: src/_pytest/cacheprovider.py [hunk 4]
```diff
@@ -5,39 +5,41 @@
...
+
+from . import paths
+from .compat import _PY2 as PY2
 
 
+@attr.s
 class Cache(object):
 
-    def __init__(self, config):
-        self.config = config
```
Comment (RonnyPfannschmidt) [inline src/_pytest/cacheprovider.py]: i realized that removing config and trace is not a good change as its "public" api

**Comment by RonnyPfannschmidt** [comment]
https://github.com/benhoyt/scandir/pull/53 for pypy on windows :/

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus is it permissible to kill off the windows pypy testing until we can get it to build scandir and/or provide a wheel?

**Pull request merged**